File: amaderpay_flutter/backend/models/Transaction.js
const mongoose = require('mongoose');

const transactionSchema = new mongoose.Schema(
  {
    trxId: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    senderPhone: {
      type: String,
      required: true,
      trim: true,
    },
    amount: {
      type: Number,
      required: true,
      min: 0,
    },
    method: {
      type: String,
      required: true,
      enum: ['BKASH', 'NAGAD', 'ROCKET'],
      uppercase: true,
    },
    rawSms: {
      type: String,
      required: true,
    },
    receivedAt: {
      type: Date,
      default: Date.now,
    },
    synced: {
      type: Boolean,
      default: true, // Synced = true once it reaches the server
    },
    deviceId: {
      type: String, // Optional: identify which device sent this
      default: 'unknown',
    },
  },
  {
    timestamps: true, // adds createdAt, updatedAt
  }
);

// Index for fast date-range queries
transactionSchema.index({ receivedAt: -1 });
transactionSchema.index({ method: 1, receivedAt: -1 });

module.exports = mongoose.model('Transaction', transactionSchema);

File: amaderpay_flutter/lib/models/transaction.dart
/// Transaction model — parsed from raw SMS
class Transaction {
  final String id;
  final String trxId;
  final String senderPhone;
  final double amount;
  final String method; // BKASH, NAGAD, ROCKET
  final String rawSms;
  final DateTime receivedAt;
  final bool synced;

  Transaction({
    required this.id,
    required this.trxId,
    required this.senderPhone,
    required this.amount,
    required this.method,
    required this.rawSms,
    required this.receivedAt,
    this.synced = false,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'trxId': trxId,
        'senderPhone': senderPhone,
        'amount': amount,
        'method': method,
        'rawSms': rawSms,
        'receivedAt': receivedAt.toIso8601String(),
        'synced': synced,
      };

  factory Transaction.fromJson(Map<String, dynamic> json) => Transaction(
        id: json['id'] ?? '',
        trxId: json['trxId'] ?? '',
        senderPhone: json['senderPhone'] ?? '',
        amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
        method: json['method'] ?? '',
        rawSms: json['rawSms'] ?? '',
        receivedAt: DateTime.tryParse(json['receivedAt'] ?? '') ?? DateTime.now(),
        synced: json['synced'] ?? false,
      );

  Transaction copyWith({bool? synced}) => Transaction(
        id: id,
        trxId: trxId,
        senderPhone: senderPhone,
        amount: amount,
        method: method,
        rawSms: rawSms,
        receivedAt: receivedAt,
        synced: synced ?? this.synced,
      );
}
